### PR TITLE
Simpler param updates with python-benedict

### PR DESCRIPTION
**Pull request by sjawhar**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

## Description of Changes
Following up on my last [comment](https://github.com/iterative/dvc/pull/4769#issuecomment-714869426) on #4769 

So I did some testing and my idea to use [benedict](https://github.com/fabiocaccamo/python-benedict) to simplify the code actually works great... ~almost~. (**Update:** figured it out) As you can see in the changes here, you can replace a lot of code with a single built-in function call. This also has the benefit of using more standard `[0]` syntax for list accessing.

~I added more tests to make sure that everything was working, then I ran into this strange behavior on a single test case:~
image

~The params are being correctly updated and written, but for some reason the experiment doesn't seem to actually run the copy step. It's strange, and I would appreciate any help chasing this down...~

I just needed to add `goo` to the list of stage params :facepalm: 

Linting is failing on changes I didn't make.

## TODO
- [x] What's up with that weird behavior?!
- [x] Run the full test suite to make sure I didn't break anything elsewhere :smile:

### CODE REVIEW

File: dvc/repo/experiments/run.py
Comment (sjawhar) [inline dvc/repo/experiments/run.py:33]: To be clear, this now returns a dict like:
```
{'path.to.key': yamlValue1, 'path[2]key': yamlValue2}
```

File: dvc/repo/experiments/__init__.py
Comment (sjawhar) [inline dvc/repo/experiments/__init__.py:313]: Not sure if you prefer local or global imports. Since this is only needed for the beta experiments feature, I figured local was appropriate.

File: tests/func/experiments/test_experiments.py
Comment (sjawhar) [inline tests/func/experiments/test_experiments.py:102]: Note the change in syntax, which I think is more in line with [your future plans](https://github.com/iterative/dvc/wiki/Parametrization#expression-syntax).

**Comment by pmrowla** [comment]
This is very cool, thanks for the PR!

Since this will add a dependency I want to discuss this internally with the rest of the DVC team first before merging it, but everything looks good to me for now.

File: tests/func/experiments/test_experiments.py
Comment (skshetry) [inline tests/func/experiments/test_experiments.py:111]: This syntax is not easy to understand/use. At a first glance, I associated those `=-` together and assumed it was removing items from the list.

Ideally, I'd like to have the following syntax:
```console
$ dvc exp run --params "foo[1]=[baz, goo]" "bar={a:2, b: 3}"
```

I understand that we split it up by comma and achieve that. We should consider using a different delimiter, maybe even space and make it look json-like.

Comment (skshetry) [reply tests/func/experiments/test_experiments.py:111]: It seems a natural progression to support appending and removing similar to [Hydra's syntax](https://hydra.cc/docs/next/advanced/override_grammar/basic/) (they use `~foo` to remove and `+foo=value` to append). 

Or, maybe even use Hydra later directly.

**Review by skshetry — APPROVED** [review]

File: tests/func/experiments/test_experiments.py
Comment (sjawhar) [reply tests/func/experiments/test_experiments.py:111]: Note that I didn't change anything regarding the syntax. This is the syntax already accepted by the yaml parser being used, which is separate from this change. I'm simply documenting what currently works.

That said, I agree that it it would be nice to be able to specify arguments that include commas, and I think that your suggestion of simply specifying a list instead of a single string would be the most straight-forward way of doing it.

**Comment by pmrowla** [comment]
Rebased this PR and created #4883 for future consideration regarding hydra

**Comment by pmrowla** [comment]
Thanks for the PR @sjawhar!

**Comment by efiop** [comment]
@sjawhar @pmrowlah Unfortunately there is no such package in conda, which means we can't use benedict as is. Is there an alternative lib that has conda package?

**Comment by pmrowla** [comment]
Hydra does have a conda package (https://github.com/iterative/dvc/issues/4883). This can probably just be reverted until we get to that issue.

**Comment by efiop** [comment]
@pmrowla Since hydra would change the syntax for this, I guess we better to do the switch before 2.0 if we indeed decide to go with it? Your call here. Could revert, or could submit benedict to conda-forge if we want (will be happy to help with sending one, we've sent a few already).

**Pull request merged**
